examples/Cassie/osc_jump/jumping_event_based_fsm.cc

```diff
@@ -0,0 +1,187 @@
+    prev_time(0) = current_time;
+    transition_flag(0) = false;
+  }
+
+  switch ((FSM_STATE)fsm_state(0)) {
+    case (BALANCE):
+      if (current_time > transition_times_[BALANCE]) {
+        fsm_state << CROUCH;
+        std::cout << "Current time: " << current_time << std::endl;
+        std::cout << "Setting fsm to CROUCH" << std::endl;
+        std::cout << "fsm: " << (FSM_STATE)fsm_state(0) << std::endl;
+        transition_flag(0) = false;
+        prev_time(0) = current_time;
+      }
+      break;
+    case (CROUCH):  // This assumes perfect knowledge about contacts
+      if (DetectGuardCondition(contact_based_
+                                   ? num_contacts == 0
+                                   : current_time > transition_times_[CROUCH],
+                               current_time, discrete_state)) {
+        state_trigger_time(0) = current_time;
+        transition_flag(0) = true;
+      }
+      if (current_time - state_trigger_time(0) >= transition_delay_ &&
+          (bool)transition_flag(0)) {
+        fsm_state << FLIGHT;
+        std::cout << "Current time: " << current_time << std::endl;
+        std::cout << "First detection time: " << state_trigger_time(0) << "\n";
+        std::cout << "Setting fsm to FLIGHT" << std::endl;
+        std::cout << "fsm: " << (FSM_STATE)fsm_state(0) << std::endl;
+        transition_flag(0) = false;
+        prev_time(0) = current_time;
+      }
+      break;
+    case (FLIGHT):
+      if (DetectGuardCondition(contact_based_
+                                   ? num_contacts != 0
+                                   : current_time > transition_times_[FLIGHT],
+                               current_time, discrete_state)) {
+        state_trigger_time(0) = current_time;
+        transition_flag(0) = true;
+      }
+      if (current_time - state_trigger_time(0) >= transition_delay_ &&
+          (bool)transition_flag(0)) {
+        fsm_state << LAND;
+        std::cout << "Current time: " << current_time << "\n";
+        std::cout << "First detection time: " << state_trigger_time(0) << "\n";
+        std::cout << "Setting fsm to LAND" << "\n";
+        std::cout << "fsm: " << (FSM_STATE)fsm_state(0) << "\n";
+        transition_flag(0) = false;
+        prev_time(0) = current_time;
+      }
+      break;
+    case (LAND):
+      break;
+  }
+
+  return EventStatus::Succeeded();
+}
+
+void JumpingEventFsm::CalcFiniteState(const Context<double>& context,
+                                      BasicVector<double>* fsm_state) const {
+  fsm_state->get_mutable_value() =
+      context.get_discrete_state().get_vector(fsm_idx_).get_value();
+}
+
+bool JumpingEventFsm::DetectGuardCondition(
+    bool guard_condition, double current_time,
+    DiscreteValues<double>* discrete_state) const {
+  auto transition_flag =
+      discrete_state->get_mutable_vector(transition_flag_idx_)
+          .get_mutable_value();
+  // Second condition is to prevent overwriting state_trigger_time
```

Comment:
I wasn't able to easily follow the FSM logic, combining this and the above code. Maybe a brief bit of documentation could help.

Reply:
done